### PR TITLE
feat: add classes for pseudo metrics with generic codomain

### PR DESCRIPTION
**Pull request by edegeltje**
This PR adds definitions for `GPseudoMetric` and `GPseudoMetricClass`.

`GPseudoMetric a b` is a structure carrying a distance function on the type `a` with a codomain `b`, along with proof that it is a pseudo metric. Furthermore, `GPseudoMetricSpace T a b` declares `T` is a type for pseudo metrics on `a` where the metric takes on values in `b`.
These notions of generic pseudo metric spaces are necessary for at least these two reasons:
1. Some metrics cannot be injectively embedded into Reals or EReals (for example the distance induced by the norm given by absolute value on hyperreal or surreal numbers). 
2. it allows the immediate use of properties of the 'actual' codomain, rather than having to frequently cast between `Real` and your actual codomain. For example, when the actual codomain of the metric is `Nat`, you can define the covering radius in a way which definitionally results in another instance of `Nat`, rather than having to prove it does.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/GPseudoMetricSpace/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+import Mathlib.Topology.GPseudoMetricSpace.IsOrderedAddCommMonoid
+
+/-!
+# General Pseudo-Metric Spaces
+
+In this file we introduce `GPseudoMetricSpace`.
+This differs from `PseudoMetricSpace` by not requiring that the codomain of the metric be `ℝ`,
+instead only requiring that it is an (additive) commutative monoid, with a linear ordering that
+respects sensible assumptions about interactions between `+` and `≤`.
+## Main Definitions
```
Comment (newell) [inline Mathlib/Topology/GPseudoMetricSpace/Basic.lean:18]:
```suggestion
respects sensible assumptions about interactions between `+` and `≤`.

## Main Definitions
```

There should be a blank line between sections and new headings.

File: Mathlib/Topology/GPseudoMetricSpace/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+inferring the intended return type without type annotations.
+Because of this, the field is set as protected, and a function
+`gdist (β :Type*) ... : α → α → β` is shadowing the name of the field.
+-/
+
+open Set Filter Bornology
+open scoped BigOperators
+
+/-- Construct a bornology from a generic distance function and metric space axioms.
+this is not declared as an instance, as it would lead to multiple instances of bornology on α,
```
Comment (newell) [inline Mathlib/Topology/GPseudoMetricSpace/Basic.lean:42]:
```suggestion
This is not declared as an instance, as it would lead to multiple instances of bornology on α,
```

File: Mathlib/Topology/GPseudoMetricSpace/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+      · refine ⟨C + C, fun a ha b hb => (gdist_triangle a x b).trans ?_⟩
+        simpa only [gdist_comm] using add_le_add (hC _ ha) (hC _ hb)
+      rcases hs with ⟨Cs, hs⟩; rcases ht with ⟨Ct, ht⟩
+      refine ⟨max Cs (gdist x y + Ct), fun z hz => hz.elim
+        (fun hz => (hs hx hz).trans (le_max_left _ _))
+        (fun hz => (gdist_triangle x y z).trans <|
+          (add_le_add le_rfl (ht hy hz)).trans (le_max_right _ _))⟩)
+    fun z => ⟨gdist z z, forall_eq.2 <| forall_eq.2 le_rfl⟩
+
+/-- the class of spaces α with a generic distance function on which return an element of type β -/
```
Comment (newell) [inline Mathlib/Topology/GPseudoMetricSpace/Basic.lean:66]:
```suggestion
/-- The class of spaces α with a generic distance function on which return an element of type β -/
```

File: Mathlib/Topology/GPseudoMetricSpace/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+@[ext]
+class GDist
+    (α : Type*) (β : Type*) [LinearOrder β] [AddCommMonoid β] [IsOrderedAddCommMonoid β] where
+  /-- the generic distance function on α to β -/
+  protected gdist : α → α → β
+
+/-- the generic distance function on a space α which returns an element of type β,
```
Comment (newell) [inline Mathlib/Topology/GPseudoMetricSpace/Basic.lean:73]:
```suggestion
/-- The generic distance function on a space α which returns an element of type β,
```

File: Mathlib/Topology/GPseudoMetricSpace/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+    (gdist_comm : ∀ x y : α, gdist x y = gdist y x)
+    (gdist_triangle : ∀ x y z : α, gdist x z ≤ gdist x y + gdist y z) : 0 ≤ gdist x y := by
+  have : 0 ≤ gdist x y + gdist x y :=
+    calc 0 = gdist x x := (gdist_self _).symm
+    _ ≤ gdist x y + gdist y x := gdist_triangle _ _ _
+    _ = gdist x y + gdist x y:= by rw [gdist_comm]
+  exact nonneg_add_self_iff.mp this
+
+/-- generic Pseudo-metric spaces
+a generic Pseudo metric space is a space endowed with some distance function `gdist : α → α → β`
```
Comment (newell) [inline Mathlib/Topology/GPseudoMetricSpace/Basic.lean:93]:
```suggestion
/-- Generic Pseudo-metric spaces

A generic Pseudo metric space is a space endowed with some distance function `gdist : α → α → β`
```

File: Mathlib/Topology/GPseudoMetricSpace/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+  funext x y; exact gdist_comm _ _
+
+theorem gdist_nonneg {x y : α} : 0 ≤ (gdist β x y) :=
+  gdist_nonneg' (gdist β) gdist_self gdist_comm gdist_triangle
+end
+
+
+namespace GMetricSpace
+section non_cancel
+variable [LinearOrder β] [AddCommMonoid β] [IsOrderedAddCommMonoid β] [GPseudoMetricSpace α β]
```
Comment (newell) [inline Mathlib/Topology/GPseudoMetricSpace/Basic.lean:200]:
```suggestion
namespace GMetricSpace

section non_cancel

variable [LinearOrder β] [AddCommMonoid β] [IsOrderedAddCommMonoid β] [GPseudoMetricSpace α β]
```

File: Mathlib/Topology/GPseudoMetricSpace/IsOrderedAddCommMonoid.lean [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+Copyright (c) 2024 Edward van de Meent. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Edward van de Meent
+-/
+import Mathlib.Algebra.Order.Monoid.Defs
+
+/-!
+# IsOrderedAddCommMonoid mixin
+
+this file defines a mixin class for combining kinds orderings and kinds of AddCommMonoid.
```
Comment (newell) [inline Mathlib/Topology/GPseudoMetricSpace/IsOrderedAddCommMonoid.lean:11]:
```suggestion
This file defines a mixin class for combining kinds orderings and kinds of AddCommMonoid.
```

File: Mathlib/Topology/GPseudoMetricSpace/IsOrderedAddCommMonoid.lean [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+Additional useful definitions:
+
+- `instance LinearOrderedAddCommMonoid α`
+given `IsOrderedAddCommMonoid α` and `LinearOrder α`
+- `instance LinearOrderedCancelAddCommMonoid α`
+given `IsOrderedCancelAddCommMonoid α` and `LinearOrder α`
+-/
+
+/--mixin for saying adding a constant does not change weak inequality-/
```
Comment (newell) [inline Mathlib/Topology/GPseudoMetricSpace/IsOrderedAddCommMonoid.lean:31]:
```suggestion
/-- Mixin for saying adding a constant does not change weak inequality-/
```

File: Mathlib/Topology/GPseudoMetricSpace/IsOrderedAddCommMonoid.lean [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+class IsOrderedAddCommMonoid (α : Type*) [PartialOrder α] [AddCommMonoid α] : Prop where
+  /- Adding a constant on the left does not change the order -/
+  protected add_le_add_left : ∀ a b : α, a ≤ b → ∀ c, c + a ≤ c + b
+
+/--mixin for saying removing the adding of a constant does not change weak inequality-/
```
Comment (newell) [inline Mathlib/Topology/GPseudoMetricSpace/IsOrderedAddCommMonoid.lean:36]:
```suggestion
/-- Mixin for saying removing the adding of a constant does not change weak inequality. -/
```

**Comment by edegeltje** [comment]
it has been pointed out to me that the `IsOrderedAddCommMonoid` already exists in a more general form, namely `CovariantClass`. i'll be refactoring my changes to use that class instead, and remove my mixins.

**Comment by grunweg** [comment]
Hello! I randomly came across your PR and noticed there is a merge conflict now. Can you merge the master branch, please?

If you do, this PR will also show up in the list of PRs awaiting review :-)

File: Mathlib/Topology/GPseudoMetric/Basic.lean [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+import Mathlib.Algebra.Order.Monoid.Lemmas
+import Mathlib.Data.Set.Lattice
+/-!
+# General Pseudo-Metrics
+
+In this file we introduce `GPseudoMetric`.
+This differs from `PseudoMetricSpace` by not requiring that the codomain of the metric be `ℝ`,
+instead only requiring that it is an (additive) commutative monoid, with a linear ordering.
+
+## Main Definitions
```
Comment (grunweg) [inline Mathlib/Topology/GPseudoMetric/Basic.lean:17]: A small comment: I think mathlib doesn't use Capitalisation In Headers much. 
```suggestion
## Main definitions

File: Mathlib/Topology/GPseudoMetric/Basic.lean [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+Copyright (c) 2024 Edward van de Meent. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Edward van de Meent
+-/
+import Mathlib.Algebra.CovariantAndContravariant
+import Mathlib.Algebra.NeZero
+import Mathlib.Algebra.Order.Monoid.Lemmas
+import Mathlib.Data.Set.Lattice
+/-!
+# General Pseudo-Metrics
```
Comment (grunweg) [inline Mathlib/Topology/GPseudoMetric/Basic.lean:11]: I'm wondering if this title can be made more specific. Is there a name for this concept in textbooks or papers? If there is, it could be useful to reference this.

To be clear: if you know this, it's wonderful - but in my personal opinion, this is not required to have your PR merged.

File: Mathlib/Topology/GPseudoMetric/Basic.lean [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+
+In this file we introduce `GPseudoMetric`.
+This differs from `PseudoMetricSpace` by not requiring that the codomain of the metric be `ℝ`,
+instead only requiring that it is an (additive) commutative monoid, with a linear ordering.
+
+## Main Definitions
+
+- `GPseudoMetric α β`: A structure containing a distance function on `α` with codomain `β`,
```
Comment (grunweg) [inline Mathlib/Topology/GPseudoMetric/Basic.lean:19]: Small grammar comment: in English, you continue in lower-case after a colon. Also in a few other places below.

(Some languages are different. In German, for instance, you use upper-case.)
```suggestion
- `GPseudoMetric α β`: a structure containing a distance function on `α` with codomain `β`,

File: Mathlib/Topology/GPseudoMetric/Basic.lean [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+which may be equal to 0 for non-equal elements. the distance function is 0 for equal elements,
+is commutative in its arguments, and satisifies the triangle inequality.
```
Comment (grunweg) [inline Mathlib/Topology/GPseudoMetric/Basic.lean:21]: Two typos: "the distance function"; "satisifies"

File: Mathlib/Topology/GPseudoMetric/Basic.lean [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+
+- `ball gdist x δ`: The set of points with distance to x strictly less than δ
+- `closedBall gdist x δ`: The set of points with distance to x less than or equal to δ
+- `sphere gdist x δ`: The set of points with distance to x equal to δ
+
+-/
+
+open Set
+
+/-- Generic pseudo metrics
```
Comment (grunweg) [inline Mathlib/Topology/GPseudoMetric/Basic.lean:34]: You wrote "pseudo-metric" above, but "pseudo metric" here. Please be consistent :-)
(To your consolation, mathlib uses both versions about equally often, so you're in good company.)

File: Mathlib/Topology/GPseudoMetric/Basic.lean [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+
+A generic pseudo metric is a distance function `gdist : α → α → β`, which is zero for
+identical elements, for which the arguments commute, and for which the triangle
+inequality holds. As opposed to a classical pseudo metric, the codomain of this distance
+function is not necessarily ℝ (or ℝ≥0∞), and as a result does not endow α with a uniform
+space.
```
Comment (grunweg) [inline Mathlib/Topology/GPseudoMetric/Basic.lean:40]: uniform space structure?

File: Mathlib/Topology/GPseudoMetric/Basic.lean [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+-/
+structure GPseudoMetric
+    (α : Type*) (β : Type*) [LinearOrder β] [AddCommMonoid β]
+    where
+  /-- A distance function on α with values in β -/
+  toFun (x y : α):β
```
Comment (grunweg) [inline Mathlib/Topology/GPseudoMetric/Basic.lean:46]:
```suggestion
  toFun (x y : α) → β

File: Mathlib/Topology/GPseudoMetric/Basic.lean [hunk 17]
```diff
@@ -0,0 +1,328 @@
...
+  gdist_self : ∀ x : α, toFun x x = 0
+  comm' : ∀ x y : α, toFun x y = toFun y x
+  triangle' : ∀ x y z : α, toFun x z ≤ toFun x y + toFun y z
+
+variable {α β :Type*} [LinearOrder β] [AddCommMonoid β] [CovariantClass β β (.+.) (.≤.)]
```
Comment (grunweg) [inline Mathlib/Topology/GPseudoMetric/Basic.lean:51]: Here and elsewhere: please put spaces around the colon, and around binary operators (i.e., `+` and `≤`).

And a blank line after the `variable`, please.

File: Mathlib/Topology/GPseudoMetric/Basic.lean [hunk 18]
```diff
@@ -0,0 +1,328 @@
...
+@[ext]
+theorem GPseudoMetric.ext {d₁ d₂ : GPseudoMetric α β} (h : d₁.toFun = d₂.toFun) : d₁ = d₂ := by
+  cases' d₁ with d _ _ _ B hB
```
Comment (grunweg) [inline Mathlib/Topology/GPseudoMetric/Basic.lean:54]: Often, for such proofs it suffices to write just
```
cases d₁
cases d₂
congr```
Does this also suffice here?

File: Mathlib/Topology/GPseudoMetric/Basic.lean [hunk 19]
```diff
@@ -0,0 +1,328 @@
...
+
+instance: GPseudoMetricClass (GPseudoMetric α β) α β where
+  gdist_self := GPseudoMetric.gdist_self
+  comm' := GPseudoMetric.comm'
+  triangle' := GPseudoMetric.triangle'
+
+
+variable {T:Type*} [FunLike T α (α → β)] [GPseudoMetricClass T α β] (gdist:T)
+
+
```
Comment (grunweg) [inline Mathlib/Topology/GPseudoMetric/Basic.lean:78]: As once above, please add spaces around the colon - and one blank line around them is enough :-)
```suggestion
variable {T : Type*} [FunLike T α (α → β)] [GPseudoMetricClass T α β] (gdist : T)

File: Mathlib/Topology/GPseudoMetric/Basic.lean [hunk 20]
```diff
@@ -0,0 +1,328 @@
...
+theorem triangle' (x y z : α) : gdist x z ≤ gdist x y + gdist y z:=
+  GPseudoMetricClass.triangle' gdist x y z
+
+theorem triangle_left (x y z : α) : gdist x y ≤ gdist z x + gdist z y := by
+  rw [comm' gdist z]; apply triangle'
+
+theorem triangle_right (x y z : α) : gdist x y ≤ gdist x z + gdist y z := by
+  rw [comm' gdist y]; apply triangle'
+
+
```
Comment (grunweg) [inline Mathlib/Topology/GPseudoMetric/Basic.lean:96]: One blank line between these theorems is enough :-)

**Review by grunweg — COMMENTED** [review]
Thank you for your contribution! I gave this a superficial look; as a not-really-expert on this area of mathlib, the file looks in broadly right shape, but a more expert reviewer will have to judge the details.

I have a number of small comments: that is perfectly normal at this point (and I like paying attention to details). Just in case: have you seen the [style guide](https://leanprover-community.github.io/contribute/style.html) already? Some things I'm saying are also repeated there, in case you want to take a look later.

File: Mathlib/Topology/GPseudoMetric/Basic.lean (hunk 11 above)
Comment (edegeltje) [reply Mathlib/Topology/GPseudoMetric/Basic.lean]: <rant>
I did some digging, and according to [this mathoverflow post](https://mathoverflow.net/a/363831) among others, these are called `[Codomain]-Metrics` or `[Codomain]-valued metrics` (and the associated spaces `[Codomain]-Metric space`, although in my definition the requirements on the codomain is somewhat weaker than described in that post. if you'd like i can change the file to reflect this? also, in the post, both `𝑅` and `Λ` are used to refer to the codomain... since the latter might be confused with conjunction, i would prefer the first, but i have no clue how common either of these uses are... but then again, [this paper](https://arxiv.org/pdf/1502.05002.pdf) does refer to (something similar to) this concept as "generalised metric space". interestingly, all of these references put more conditions on the codomain than i expected... in particular, they expect the ordering to be co- and contravariant compared to addition...
</rant>
tldr: the version in the PR is close to an existing concept, which gets described as "generalised metric space". such spaces are called `R`-Metric spaces or `R`-valued metric spaces where `R` is the codomain, although capital Lambda is also sometimes used. The concept in this PR is slightly weaker with regards to the assumptions on the codomain.

Comment (edegeltje) [reply Mathlib/Topology/GPseudoMetric/Basic.lean]: also, according to [this other paper](http://www.tac.mta.ca/tac/reprints/articles/1/tr1.pdf), there are ties with category theory as well

File: Mathlib/Topology/GPseudoMetric/Basic.lean [hunk 21]
```diff
@@ -0,0 +1,332 @@
...
+- `GPseudoMetric α β`: a structure containing a distance function on `α` with codomain `β`,
+which may be equal to `0` for non-equal elements. The distance function is `0` for equal elements,
+is commutative in its arguments, and satisfies the triangle inequality.
+- `GPseudoMetricClass α β`: the class of types of generic pseudo-metrics on `α` to `β`.
+
+Additional useful definitions:
+
+- `ball gdist x δ`: the set of points with distance to `x` strictly less than `δ`
+- `closedBall gdist x δ`: the set of points with distance to `x` less than or equal to `δ`
+- `sphere gdist x δ`: the set of points with distance to `x` equal to `δ`
```
Comment (YaelDillies) [inline Mathlib/Topology/GPseudoMetric/Basic.lean:29]: Please fully qualify names so that the links work in the docs
```suggestion
- `GMetric.ball gdist x δ`: the set of points with distance to `x` strictly less than `δ`
- `GMetric.closedBall gdist x δ`: the set of points with distance to `x` less than or equal to `δ`
- `GMetric.sphere gdist x δ`: the set of points with distance to `x` equal to `δ`
```

File: Mathlib/Topology/GPseudoMetric/Basic.lean [hunk 22]
```diff
@@ -0,0 +1,332 @@
...
+open Set
+
+/-- Generic pseudo-metrics
+
+A generic pseudo-metric is a distance function `gdist : α → α → β`, which is zero for identical
+elements, for which the arguments commute, and for which the triangle inequality holds.
+As opposed to a classical pseudo-metric, the codomain of this distance function is not
+necessarily `ℝ` (or `ℝ≥0∞`), and as a result does not endow `α` with a uniform space structure.
+-/
+
```
Comment (YaelDillies) [inline Mathlib/Topology/GPseudoMetric/Basic.lean:41]:
```suggestion
```

File: Mathlib/Topology/GPseudoMetric/Basic.lean
Comment (YaelDillies) [inline Mathlib/Topology/GPseudoMetric/Basic.lean:1]: You are not talking about pseudo-metrics only, so you might as well call your file `Topology.GMetric.Basic` or `Topology.GMetric`

**Review by YaelDillies — COMMENTED** [review]
Is there a specific reason why you are using the "type of metrics" + "hom class" design (which we use for morphisms in algebra, eg `MonoidHom` + `MonoidHomClass`) rather than the "canonical metric" design (`PseudoMetricSpace` or `MetricSpace`).

If so, can you underline that reason in the module doc. If not, can you drop that approach and use the canonical metric approach instead? (please let me know if you don't know what needs to be done to switch)

**Comment by riccardobrasca** [comment]
@edegeltje Are you still working on this PR? I think there are design decisions we are still not sure about, see [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Packings.2C.20coverings.2C.20nets.2C.20uniform.20discretesness.20et.2E.20al.2E). Maybe you can discuss the PR there? Thanks!

**Comment by edegeltje** [comment]
i have for now abandoned working on this. although i do believe creating this is the correct direction compared to the current class-based unbundled approach, i don't know how to make progress with this...

**Comment by semorrison** [comment]
Discussion amongst reviewers seems to have concluded that a design more like
```
class NatDist (α : Type*) [PseudoEMetricSpace α] where
  natDist : α → α → ℕ
  natCast_natDist (a b) : natDist a b = edist a b
```
is the right path forward here.

Given your comment above, I'm going to close this for now, but hopefully we will see further progress on this when you have time again!
